@xctan Looks like something does not work correctly with symbols naming:
```
[ 14%] Building CXX object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/llamafile/sgemm.cpp.o
cd /opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/build/ggml/src && /opt/local/bin/g++-mp-14 -DACCELERATE_LAPACK_ILP64 -DACCELERATE_NEW_LAPACK -DGGML_BACKEND_BUILD -DGGML_BACKEND_SHARED -DGGML_SCHED_MAX_COPIES=4 -DGGML_SHARED -DGGML_USE_ACCELERATE -DGGML_USE_CPU_REPACK -DGGML_USE_LLAMAFILE -DGGML_USE_OPENMP -D_DARWIN_C_SOURCE -D_XOPEN_SOURCE=600 -Dggml_cpu_EXPORTS -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/.. -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/. -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/ggml-cpu -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/../include -F//System/Library/Frameworks -pipe -Os -fpermissive -DNDEBUG -isystem/opt/local/include/LegacySupport -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=gnu++17 -flto=auto -arch ppc -mmacosx-version-min=10.6 -fPIC -Wmissing-declarations -Wmissing-noreturn -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wno-array-bounds -Wextra-semi -fopenmp -MD -MT ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/llamafile/sgemm.cpp.o -MF CMakeFiles/ggml-cpu.dir/ggml-cpu/llamafile/sgemm.cpp.o.d -o CMakeFiles/ggml-cpu.dir/ggml-cpu/llamafile/sgemm.cpp.o -c /opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/ggml-cpu/llamafile/sgemm.cpp
[ 14%] Building C object ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/powerpc/quants.c.o
cd /opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/build/ggml/src && /opt/local/bin/gcc-mp-14 -DACCELERATE_LAPACK_ILP64 -DACCELERATE_NEW_LAPACK -DGGML_BACKEND_BUILD -DGGML_BACKEND_SHARED -DGGML_SCHED_MAX_COPIES=4 -DGGML_SHARED -DGGML_USE_ACCELERATE -DGGML_USE_CPU_REPACK -DGGML_USE_LLAMAFILE -DGGML_USE_OPENMP -D_DARWIN_C_SOURCE -D_XOPEN_SOURCE=600 -Dggml_cpu_EXPORTS -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/.. -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/. -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/ggml-cpu -I/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/../include -F//System/Library/Frameworks -pipe -Os -DNDEBUG -isystem/opt/local/include/LegacySupport -I/opt/local/include -std=gnu11 -flto=auto -arch ppc -mmacosx-version-min=10.6 -fPIC -Wshadow -Wstrict-prototypes -Wpointer-arith -Wmissing-prototypes -Werror=implicit-int -Werror=implicit-function-declaration -Wall -Wextra -Wpedantic -Wcast-qual -Wno-unused-function -Wdouble-promotion -fopenmp -MD -MT ggml/src/CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/powerpc/quants.c.o -MF CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/powerpc/quants.c.o.d -o CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/powerpc/quants.c.o -c /opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/llama.cpp-5662/ggml/src/ggml-cpu/arch/powerpc/quants.c
[ 14%] Linking CXX shared library ../../../bin/libggml-blas.dylib
cd /opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/build/ggml/src/ggml-blas && /opt/local/bin/cmake -E cmake_link_script CMakeFiles/ggml-blas.dir/link.txt --verbose=ON
/opt/local/bin/g++-mp-14 -pipe -Os -fpermissive -DNDEBUG -isystem/opt/local/include/LegacySupport -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -flto=auto -arch ppc -mmacosx-version-min=10.6 -dynamiclib -Wl,-headerpad_max_install_names -Wl,-headerpad_max_install_names -L/opt/local/lib -lMacportsLegacySupport -o ../../../bin/libggml-blas.dylib -install_name @rpath/libggml-blas.dylib "CMakeFiles/ggml-blas.dir/ggml-blas.cpp.o"  -Wl,-rpath,/opt/local/lib ../../../bin/libggml-base.dylib -framework Accelerate
make[2]: Leaving directory `/opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/build'
[ 14%] Built target ggml-blas
[ 14%] Linking CXX shared library ../../bin/libggml-cpu.dylib
cd /opt/local/var/macports/build/_opt_local_ppcports_llm_llama.cpp/llama.cpp/work/build/ggml/src && /opt/local/bin/cmake -E cmake_link_script CMakeFiles/ggml-cpu.dir/link.txt --verbose=ON
Undefined symbols for architecture ppc:
  "_ggml_gemm_iq4_nl_4x4_q8_0", referenced from:
      void ggml::cpu::repack::gemm<block_iq4_nl, 4ll, 4ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_iq4_nl, 4ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemm_iq4_nl_4x4_q8_0_generic)
  "_ggml_gemm_q4_0_4x4_q8_0", referenced from:
      void ggml::cpu::repack::gemm<block_q4_0, 4ll, 4ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 4ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemm_q4_0_4x4_q8_0_generic)
  "_ggml_gemm_q4_0_4x8_q8_0", referenced from:
      void ggml::cpu::repack::gemm<block_q4_0, 8ll, 4ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 8ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemm_q4_0_4x8_q8_0_generic)
  "_ggml_gemm_q4_0_8x8_q8_0", referenced from:
      void ggml::cpu::repack::gemm<block_q4_0, 8ll, 8ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 8ll, 8ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemm_q4_0_8x8_q8_0_generic)
  "_ggml_gemm_q4_K_8x8_q8_K", referenced from:
      void ggml::cpu::repack::gemm<block_q4_K, 8ll, 8ll, (ggml_type)15>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_K, 8ll, 8ll, (ggml_type)15>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemm_q4_K_8x8_q8_K_generic)
  "_ggml_gemv_iq4_nl_4x4_q8_0", referenced from:
      void ggml::cpu::repack::gemv<block_iq4_nl, 4ll, 4ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_iq4_nl, 4ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemv_iq4_nl_4x4_q8_0_generic)
  "_ggml_gemv_q4_0_4x4_q8_0", referenced from:
      void ggml::cpu::repack::gemv<block_q4_0, 4ll, 4ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 4ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemv_q4_0_4x4_q8_0_generic)
  "_ggml_gemv_q4_0_4x8_q8_0", referenced from:
      void ggml::cpu::repack::gemv<block_q4_0, 8ll, 4ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 8ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemv_q4_0_4x8_q8_0_generic)
  "_ggml_gemv_q4_0_8x8_q8_0", referenced from:
      void ggml::cpu::repack::gemv<block_q4_0, 8ll, 8ll, (ggml_type)8>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 8ll, 8ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemv_q4_0_8x8_q8_0_generic)
  "_ggml_gemv_q4_K_8x8_q8_K", referenced from:
      void ggml::cpu::repack::gemv<block_q4_K, 8ll, 8ll, (ggml_type)15>(int, float*, unsigned long, void const*, void const*, int, int) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_K, 8ll, 8ll, (ggml_type)15>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_gemv_q4_K_8x8_q8_K_generic)
  "_ggml_quantize_mat_q8_0_4x4", referenced from:
      void ggml_quantize_mat_t<4ll, (ggml_type)8>(float const*, void*, long long, long long) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 4ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_iq4_nl, 4ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_quantize_mat_q8_0_4x4_generic)
  "_ggml_quantize_mat_q8_0_4x8", referenced from:
      void ggml_quantize_mat_t<8ll, (ggml_type)8>(float const*, void*, long long, long long) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 8ll, 4ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_0, 8ll, 8ll, (ggml_type)8>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_quantize_mat_q8_0_4x8_generic)
  "_ggml_quantize_mat_q8_K_4x8", referenced from:
      void ggml_quantize_mat_t<8ll, (ggml_type)15>(float const*, void*, long long, long long) in repack.cpp.o
      ggml::cpu::repack::tensor_traits<block_q4_K, 8ll, 8ll, (ggml_type)15>::compute_forward(ggml_compute_params*, ggml_tensor*) in repack.cpp.o
     (maybe you meant: _ggml_quantize_mat_q8_K_4x8_generic)
  "_ggml_vec_dot_iq1_m_q8_K", referenced from:
      _type_traits_cpu in ggml-cpu.c.o
     (maybe you meant: _ggml_vec_dot_iq1_m_q8_K_generic)
  "_ggml_vec_dot_tq1_0_q8_K", referenced from:
      _type_traits_cpu in ggml-cpu.c.o
     (maybe you meant: _ggml_vec_dot_tq1_0_q8_K_generic)
  "_ggml_vec_dot_tq2_0_q8_K", referenced from:
      _type_traits_cpu in ggml-cpu.c.o
     (maybe you meant: _ggml_vec_dot_tq2_0_q8_K_generic)
  "_quantize_row_q8_K", referenced from:
      _type_traits_cpu in ggml-cpu.c.o
     (maybe you meant: _quantize_row_q8_K_generic)
ld: symbol(s) not found for architecture ppc
collect2: error: ld returned 1 exit status
/opt/local/bin/g++-mp-14 -pipe -Os -fpermissive -DNDEBUG -isystem/opt/local/include/LegacySupport -I/opt/local/include -D_GLIBCXX_USE_CXX11_ABI=0 -flto=auto -arch ppc -mmacosx-version-min=10.6 -dynamiclib -Wl,-headerpad_max_install_names -Wl,-headerpad_max_install_names -L/opt/local/lib -lMacportsLegacySupport -o ../../bin/libggml-cpu.dylib -install_name @rpath/libggml-cpu.dylib "CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu.c.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/ggml-cpu.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/repack.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/hbm.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/quants.c.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/traits.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/amx/amx.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/amx/mmq.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/binary-ops.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/unary-ops.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/vec.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/ops.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/llamafile/sgemm.cpp.o" "CMakeFiles/ggml-cpu.dir/ggml-cpu/arch/powerpc/quants.c.o"  -Wl,-rpath,/opt/local/lib ../../bin/libggml-base.dylib -framework Accelerate /opt/local/lib/gcc14/libgomp.dylib
```

--- a/ggml/cmake/common.cmake
+++ b/ggml/cmake/common.cmake
@@ -36,9 +36,7 @@ function(ggml_get_system_arch)
             (NOT CMAKE_OSX_ARCHITECTURES AND NOT CMAKE_GENERATOR_PLATFORM_LWR AND
             CMAKE_SYSTEM_PROCESSOR MATCHES "^(x86_64|i686|AMD64|amd64)$"))
         set(GGML_SYSTEM_ARCH "x86" PARENT_SCOPE)
-    elseif ("${CMAKE_SYSTEM_PROCESSOR} " STREQUAL "ppc64le " OR
-            "${CMAKE_SYSTEM_PROCESSOR} " STREQUAL "ppc64 " OR
-            "${CMAKE_SYSTEM_PROCESSOR} " STREQUAL "powerpc ")
+    elseif (${CMAKE_SYSTEM_PROCESSOR} MATCHES "ppc|power")
         set(GGML_SYSTEM_ARCH "PowerPC" PARENT_SCOPE)
     elseif (${CMAKE_SYSTEM_PROCESSOR} MATCHES "loongarch64")
         set(GGML_SYSTEM_ARCH "loongarch64"  PARENT_SCOPE)

--- a/ggml/src/ggml-cpu/apple-fallback.h
+++ b/ggml/src/ggml-cpu/apple-fallback.h
@@ -64,7 +64,7 @@
 #define ggml_gemm_q4_0_4x4_q8_0_generic ggml_gemm_q4_0_4x4_q8_0
 #define ggml_gemm_q4_0_4x8_q8_0_generic ggml_gemm_q4_0_4x8_q8_0
 #define ggml_gemm_iq4_nl_4x4_q8_0_generic ggml_gemm_iq4_nl_4x4_q8_0
-#elif defined(__powerpc__)
+#elif defined(__PPC__)
 // quants.c
 #define quantize_row_q8_K_generic quantize_row_q8_K
 #define ggml_vec_dot_tq1_0_q8_K_generic ggml_vec_dot_tq1_0_q8_K

--- a/ggml/src/ggml-cpu/apple-fallback.h
+++ b/ggml/src/ggml-cpu/apple-fallback.h
@@ -64,7 +64,8 @@
 #define ggml_gemm_q4_0_4x4_q8_0_generic ggml_gemm_q4_0_4x4_q8_0
 #define ggml_gemm_q4_0_4x8_q8_0_generic ggml_gemm_q4_0_4x8_q8_0
 #define ggml_gemm_iq4_nl_4x4_q8_0_generic ggml_gemm_iq4_nl_4x4_q8_0
-#elif defined(__PPC__)
+#elif defined(__POWERPC__)
+// ref: https://github.com/ggml-org/llama.cpp/pull/14146#issuecomment-2972561679
 // quants.c
 #define quantize_row_q8_K_generic quantize_row_q8_K
 #define ggml_vec_dot_tq1_0_q8_K_generic ggml_vec_dot_tq1_0_q8_K

